ls1intum/Artemis
# `Programming exercises`: Allow to import exercises from file

### Checklist
#### General


#### Server
#### Client
#### Changes affecting Programming Exercises

### Motivation and Context
Currently, one can only import programming exercises from the same instance. This makes it impossible to share exercises between different universities. This PR adds the option to import exercises from a zip file that can be generated using the download exercise button (available for instructors).

### Description
On the client side I added three new components to extend the import-functionality (see screencast)
The client unzips the top level of the zip file to extract the exercise details from a JSON file and then opens the `programming-exercise-update-component`. When the user clicks import a call to the new server endpoint `/api/programming-exercises-from-file` with the Programming exercise object and the zip file . In this resource method, the `ProgrammingExerciseImportService` is used to import the exercise. Importing involves the following steps:
* Extracting all levels of the zip file
* set up programming exercise in the VCS and CI system
* delete the template that is added when the exercise is created initially
* copy over the content from the extracted repository zip archives into the repositories and push it.

**Known limitations**
* Categories, team size and submission policy cannot be imported because they are missing in the exported JSON file. (out of scope for this PR)

### Steps for Testing

Prerequisites:
- 1 Instructor as only instructors can export the exercise

1. Log in to Artemis
2. Navigate to Course Administration
3. Create a programming exercise of a supported type/language
4. Export it
5. Possibly make some changes to the source code
6. Export the exercise
7. Go to Import Programming Exercise and choose From File
8. Upload the previously exported ZIP file
9. Change something in the upcoming dialog and import it
10. Make sure the correct content is in the repository and the import is successful.

#### Exam Mode Testing
Prerequisites:
- 1 Instructor as only instructors can export the exercise

1. Log in to Artemis
2. Navigate to Course Administration
3. Create a programming exercise of a supported type/language
4. Export it
5. Possibly make some changes to the source code
6. Export the exercise
7. Go to exams
8. Go to exercise groups
9. Go to Import programming exercise
10. Upload the previously exported ZIP file
11. Change something in the upcoming dialog and import it
12. Make sure the correct content is in the repository and the import is successful.

### Review Progress






#### Code Review
#### Manual Tests
#### Exam Mode Test

### Test Coverage




| Class/File | Branch | Line | Comment |
|------------|-------:|-----:|-----:|
| ProgrammingExerciseExportImportResource.java | 81% | 96% |
| ProgrammingExerciseImportFromFileService.java  | 85% | 93% |
| ProgrammingExercisService.java | 82% | 93% |
| ProgrammingExerciseResource.java | 76% | 89% |
| programming-exercise-update-component |74.44% |80.7%| (my changes are fully covered) |
| exercise-import-from-file-component.ts | 80% | 95.54% |
| exercise-import-tabs-component.ts | 100% | 100% |
| programming-exercise.service.ts | 61.9% | 61.29% | (my changes are fully covered) |
| exercise-import-wrapper-component.ts  | 87.5% | 100% |
| RespositoryService.java | 88% | 92% |  
| ZipFileService.java | 68% | 95% |



### Screenshots



https://user-images.githubusercontent.com/84102468/227810296-b1c7d2df-3867-46a7-852d-8f6b799fadb5.mp4
